File: src/infrastructure/amazon/downloader.js
/* exported Downloader */

class Downloader{
  constructor(authToken){
    this.SP_API_URL = "https://sellingpartnerapi-fe.amazon.com/inbound/fba/2024-03-20/items/labels";
    this.MAX_QUANTITY_PER_REQUEST = 999;
    this.authToken = authToken;
    this.options = {
      method: 'post',
      muteHttpExceptions : true,
      headers: {
        "Accept" : "application/json",
        "x-amz-access-token": authToken,
        "Content-Type" : "application/json"
      }
    };
  }

  downloadLabels(skuNums, fileName){
    const chunks = this._splitByQuantityLimit(skuNums);
    const folder = DriveApp.getFolderById("1ymbSzyiawRaREUwwaNYp4OzoGEOBgDNp");
    const totalChunks = chunks.length;

    if (totalChunks === 1) {
      return this._downloadSingleBatch(chunks[0], fileName, folder);
    }

    return this._downloadMultipleBatches(chunks, fileName, folder);
  }

  _downloadSingleBatch(skuNums, fileName, folder) {
    const responseJson = this._fetchLabels(skuNums);
    const pdfBlob = this._downloadPdfBlob(responseJson);
    const file = folder.createFile(pdfBlob.setName(fileName + '.pdf'));
    const downloadUrl = `https://drive.google.com/uc?export=download&id=${file.getId()}`;

    return { url: downloadUrl, responseData: responseJson };
  }

  _downloadMultipleBatches(chunks, fileName, folder) {
    const urls = [];
    let lastResponseJson = null;

    for (let i = 0; i < chunks.length; i++) {
      console.log(`ラベル分割ダウンロード: ${i + 1}/${chunks.length}`);
      const responseJson = this._fetchLabels(chunks[i]);
      const pdfBlob = this._downloadPdfBlob(responseJson);
      const partName = `${fileName}_part${i + 1}`;
      const file = folder.createFile(pdfBlob.setName(partName + '.pdf'));
      urls.push(`https://drive.google.com/uc?export=download&id=${file.getId()}`);
      lastResponseJson = responseJson;
    }

    console.log(`ラベル分割ダウンロード完了: ${chunks.length}件のPDFを作成しました`);
    return { url: urls[0], urls: urls, responseData: lastResponseJson };
  }

  _fetchLabels(skuNums) {
    const payload = {
      labelType: 'STANDARD_FORMAT',
      marketplaceId: 'A1VC38T7YXB528',
      mskuQuantities: skuNums,
      localeCode: "ja_JP",
      pageType: 'A4_40_52x29'
    };

    const options = Object.assign({}, this.options, {
      payload: JSON.stringify(payload)
    });

    const response = UrlFetchApp.fetch(this.SP_API_URL, options);
    const responseJson = JSON.parse(response.getContentText());
    console.log('downloadLabels response:', responseJson);

    if (responseJson.errors && responseJson.errors.length > 0) {
      const errorMessages = responseJson.errors.map(e => `${e.code}: ${e.message}`).join('; ');
      throw new Error(`SP-API ラベル取得エラー: ${errorMessages}`);
    }

    if (!responseJson.documentDownloads || responseJson.documentDownloads.length === 0) {
      throw new Error('SP-API レスポンスにダウンロードURLが含まれていません');
    }

    return responseJson;
  }

  _downloadPdfBlob(responseJson) {
    const fileURI = responseJson.documentDownloads[0].uri;
    const fileResponse = UrlFetchApp.fetch(fileURI, {method:"GET"});
    return fileResponse.getBlob();
  }

  _splitByQuantityLimit(skuNums) {
    const maxQty = this.MAX_QUANTITY_PER_REQUEST;
    const totalQuantity = skuNums.reduce((sum, item) => sum + item.quantity, 0);

    if (totalQuantity <= 15000) {
      return [skuNums];
    }

    const chunks = [];
    const currentChunk = [];
    let expanded = [];

    for (const item of skuNums) {
      if (item.quantity <= maxQty) {
        expanded.push(item);
      } else {
        let remaining = item.quantity;
        while (remaining > 0) {
          const qty = Math.min(remaining, maxQty);
          expanded.push({ msku: item.msku, quantity: qty });
          remaining -= qty;
        }
      }
    }

    let chunkTotal = 0;
    for (const item of expanded) {
      if (chunkTotal + item.quantity > maxQty && currentChunk.length > 0) {
        chunks.push([...currentChunk]);
        currentChunk.length = 0;
        chunkTotal = 0;
      }
      currentChunk.push(item);
      chunkTotal += item.quantity;
    }

    if (currentChunk.length > 0) {
      chunks.push(currentChunk);
    }

    console.log(`ラベル数量が上限(${maxQty})を超えるため ${chunks.length} 回に分割します`);
    return chunks;
  }
}

File: python/src/infrastructure/amazon/downloader.py
from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

import httpx

logger = logging.getLogger(__name__)

SP_API_LABELS_URL = "https://sellingpartnerapi-fe.amazon.com/inbound/fba/2024-03-20/items/labels"
MAX_QUANTITY_PER_REQUEST = 999


class Downloader:
    def __init__(self, auth_token: str, save_dir: Path) -> None:
        self._auth_token = auth_token
        self._save_dir = save_dir
        self._headers = {
            "Accept": "application/json",
            "x-amz-access-token": auth_token,
            "Content-Type": "application/json",
        }

    def download_labels(self, sku_nums: list[dict[str, Any]], file_name: str) -> dict[str, Any]:
        chunks = self._split_by_quantity_limit(sku_nums)
        if len(chunks) == 1:
            return self._download_single_batch(chunks[0], file_name)
        return self._download_multiple_batches(chunks, file_name)

    def _download_single_batch(self, sku_nums: list[dict[str, Any]], file_name: str) -> dict[str, Any]:
        response_json = self._fetch_labels(sku_nums)
        pdf_bytes = self._download_pdf_bytes(response_json)
        path = self._save_to_local(pdf_bytes, f"{file_name}.pdf")
        return {"path": str(path), "response_data": response_json}

    def _download_multiple_batches(self, chunks: list[list[dict[str, Any]]], file_name: str) -> dict[str, Any]:
        paths: list[str] = []
        last_response: dict[str, Any] = {}
        for i, chunk in enumerate(chunks):
            logger.info("ラベル分割ダウンロード: %d/%d", i + 1, len(chunks))
            response_json = self._fetch_labels(chunk)
            pdf_bytes = self._download_pdf_bytes(response_json)
            path = self._save_to_local(pdf_bytes, f"{file_name}_part{i + 1}.pdf")
            paths.append(str(path))
            last_response = response_json
        logger.info("ラベル分割ダウンロード完了: %d件のPDFを作成しました", len(chunks))
        return {"path": paths[0], "paths": paths, "response_data": last_response}

    def _fetch_labels(self, sku_nums: list[dict[str, Any]]) -> dict[str, Any]:
        payload = {
            "labelType": "STANDARD_FORMAT",
            "marketplaceId": "A1VC38T7YXB528",
            "mskuQuantities": sku_nums,
            "localeCode": "ja_JP",
            "pageType": "A4_40_52x29",
        }
        response = httpx.post(SP_API_LABELS_URL, json=payload, headers=self._headers, timeout=30.0)
        response_json = response.json()
        if response_json.get("errors"):
            messages = "; ".join(f'{e["code"]}: {e["message"]}' for e in response_json["errors"])
            raise RuntimeError(f"SP-API ラベル取得エラー: {messages}")
        if not response_json.get("documentDownloads"):
            raise RuntimeError("SP-API レスポンスにダウンロードURLが含まれていません")
        return response_json

    def _download_pdf_bytes(self, response_json: dict[str, Any]) -> bytes:
        file_uri = response_json["documentDownloads"][0]["uri"]
        response = httpx.get(file_uri, timeout=60.0)
        response.raise_for_status()
        return response.content

    def _save_to_local(self, pdf_bytes: bytes, file_name: str) -> Path:
        self._save_dir.mkdir(parents=True, exist_ok=True)
        file_path = self._save_dir / file_name
        file_path.write_bytes(pdf_bytes)
        logger.info("ラベルPDF保存: %s", file_path)
        return file_path

    def _split_by_quantity_limit(self, sku_nums: list[dict[str, Any]]) -> list[list[dict[str, Any]]]:
        SPLIT_THRESHOLD = 15000
        CHUNK_LIMIT = 10000
        total_quantity = sum(item["quantity"] for item in sku_nums)
        if total_quantity <= SPLIT_THRESHOLD:
            return [sku_nums]
        chunks: list[list[dict[str, Any]]] = []
        current_chunk: list[dict[str, Any]] = []
        chunk_total = 0
        for item in sku_nums:
            if chunk_total + item["quantity"] > CHUNK_LIMIT and current_chunk:
                chunks.append(current_chunk)
                current_chunk = []
                chunk_total = 0
            current_chunk.append(item)
            chunk_total += item["quantity"]
        if current_chunk:
            chunks.append(current_chunk)
        return chunks
